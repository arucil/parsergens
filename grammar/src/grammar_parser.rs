
pub mod lex;
pub mod ast;

use lalrpop_util::{lalrpop_mod, ParseError};

lalrpop_mod!(pub grammar, "/grammar_parser/grammar.rs");

pub fn parse(
  input: &str
) -> Result<ast::Grammar, ParseError<usize, lex::Token, lex::LexError>> {
  let lexer = lex::Lexer::new(input);
  grammar::DocumentParser::new().parse(input, lexer)
}
