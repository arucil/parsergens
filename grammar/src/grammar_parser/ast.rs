use std::ops::Range;

pub type Grammar = Vec<Spanned<Decl>>;

#[derive(Debug)]
pub enum Decl {
  Token(TokenDecl),
  Start(StartDecl),
  Rule(RuleDecl),
  Skip(SkipDecl),
}

#[derive(Debug)]
pub struct TokenDecl {
  pub name: Spanned<String>,
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub enum TokenPattern {
  Regex(Spanned<(String, Regex)>),
  String(Spanned<(String, Regex)>),
}

#[derive(Debug)]
pub enum Regex {
  Any,
  Char(char),
  CharSet(Vec<Range<char>>),
  CharClass(CharClassKind),
  Or(Box<Regex>, Box<Regex>),
  Optional(Box<Regex>),
  Many(Box<Regex>),
  Many1(Box<Regex>),
  Seq(Vec<Regex>),
}

#[derive(Debug)]
pub enum CharClassKind {
  Digit,
  Word,
}

#[derive(Debug)]
pub struct SkipDecl {
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub struct StartDecl {
  pub name: Spanned<String>,
}

#[derive(Debug)]
pub struct RuleDecl {
  pub name: Spanned<String>,
  pub alts: Vec<Spanned<RuleAlt>>,
}

#[derive(Debug)]
pub enum RuleAlt {
  Epsilon,
  Terms(Vec<Spanned<String>>),
}

#[derive(Debug)]
pub struct Spanned<T>(pub Range<usize>, pub T);