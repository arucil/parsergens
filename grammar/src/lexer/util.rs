
pub fn find_char_interval(char: u32, char_intervals: &[u32]) -> u32 {
  let mut lo = 0;
  let mut hi = char_intervals.len();

  while lo < hi {
    let mid = (lo + hi) / 2;
    if char_intervals[mid] > char {
      hi = mid;
    } else {
      lo = mid + 1;
    }
  }

  lo as u32 - 1
}
