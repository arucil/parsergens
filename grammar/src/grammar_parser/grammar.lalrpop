use lalrpop_util::ParseError;
use super::ast::*;
use super::lex;
use super::UserParseError;
use super::regex;

#[LALR]
grammar<'input>(input: &'input str);

pub Document: Grammar = (Separator <decl>)*
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned((l, r), Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned((l, r), Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned((l, r), Decl::Token(decl))),
  <l:@L> <decl:ext_token_decl> <r:@L> => Some(Spanned((l, r), Decl::ExtToken(decl))),
  <l:@L> <decl:skip_decl> <r:@L> => Some(Spanned((l, r), Decl::Skip(decl))),
  <l:@L> <decl:user_decl> <r:@L> => Some(Spanned((l, r), Decl::User(decl))),
  <l:@L> <decl:assoc_decl> <r:@L> => Some(Spanned((l, r), Decl::Assoc(decl))),
  <l:@L> <decl:state_decl> <r:@L> => Some(Spanned((l, r), Decl::State(decl))),
};

rule_decl: RuleDecl =
  <name:ident> <ty:(Colon <code_block>)?> Assign <alt1:production> <mut alts:(Or <production>)*>
  =>? {
    alts.insert(0, alt1);
    if alts.iter().any(|alt| alt.1.action.is_some()) && ty.is_none() {
      Err(ParseError::User {
        error: UserParseError::MissingType(name.0),
      })
    } else {
      Ok(RuleDecl {
        name,
        ty,
        alts,
      })
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:term+> <prec:(Percent Prec <ident>)?> <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Terms(terms),
        prec,
        action,
      }
    ),
  <l:@L> epsilon <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Epsilon,
        prec: None,
        action,
      }
    ),
};

epsilon: () = LParen RParen;

term: Term = {
  ident => Term::Symbol(<>),
  LParen <terms:term+> RParen QuestionMark => Term::Optional(terms),
  LParen <terms:term+> RParen Asterisk => Term::Many(terms),
  LParen <terms:term+> RParen Plus => Term::Many1(terms),
}

start_decl: StartDecl = Percent Start <name:ident>
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <pattern:token_pattern>
  => {
    TokenDecl {
      name,
      pattern,
    }
  };

ext_token_decl: ExternalTokenDecl = Percent Token <name:ident>
  => {
    ExternalTokenDecl {
      name,
    }
  };

skip_decl: SkipDecl = Percent Skip <pattern:token_pattern>
  => {
    SkipDecl {
      pattern,
    }
  };

user_decl: UserDecl = Percent User <code:code_block>
  => {
    UserDecl {
      code,
    }
  };

assoc_decl: AssocDecl = Percent <assoc:assoc> <names:ident+>
  => {
    AssocDecl {
      names,
      assoc,
    }
  };

assoc: Spanned<Assoc> = {
  <l:@L> LeftAssoc <r:@L> => Spanned((l, r), Assoc::LeftAssoc),
  <l:@L> RightAssoc <r:@L> => Spanned((l, r), Assoc::RightAssoc),
  <l:@L> NonAssoc <r:@L> => Spanned((l, r), Assoc::NonAssoc),
}

state_decl: StateDecl = Percent State <state:code_block>
  => {
    StateDecl {
      state,
    }
  };

token_pattern: TokenPattern = {
  <regex:regex> =>? regex::parse_regex(
    &regex.1[1..regex.1.len() - 1],
    regex.0 .0 + 1
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::Regex,
      source: regex,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error),
    }),
  <str:string> =>? regex::parse_string_literal(
    &str.1[1..str.1.len() - 1], str.0 .0
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::String,
      source: str,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error)
    }),
}

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned((l, r), x.text.to_owned());

ident = spanned<Ident>;
string = spanned<String>;
regex = spanned<Regex>;
code_block: Spanned<String> = <spanned<CodeBlock>> =>
  Spanned(<>.0, <>.1.trim_start_matches('{').trim_end_matches('}').trim().to_owned());

extern {
  type Location = usize;
  type Error = UserParseError;

  enum lex::Token<'input> {
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Skip => lex::Token { kind: lex::TokenKind::Skip, .. },
    User => lex::Token { kind: lex::TokenKind::User, .. },
    LeftAssoc => lex::Token { kind: lex::TokenKind::LeftAssoc, .. },
    RightAssoc => lex::Token { kind: lex::TokenKind::RightAssoc, .. },
    NonAssoc => lex::Token { kind: lex::TokenKind::NonAssoc, .. },
    Prec => lex::Token { kind: lex::TokenKind::Prec, .. },
    State => lex::Token { kind: lex::TokenKind::State, .. },

    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },
    String => lex::Token { kind: lex::TokenKind::String, .. },
    CodeBlock => lex::Token { kind: lex::TokenKind::CodeBlock, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    LParen => lex::Token { kind: lex::TokenKind::LParen, .. },
    RParen => lex::Token { kind: lex::TokenKind::RParen, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },
    Colon => lex::Token { kind: lex::TokenKind::Colon, .. },
    Asterisk => lex::Token { kind: lex::TokenKind::Asterisk, .. },
    QuestionMark => lex::Token { kind: lex::TokenKind::QuestionMark, .. },
    Plus => lex::Token { kind: lex::TokenKind::Plus, .. },
    Arrow => lex::Token { kind: lex::TokenKind::Arrow, .. },

    Enter => lex::Token { kind: lex::TokenKind::Enter, .. },
    Leave => lex::Token { kind: lex::TokenKind::Leave, .. },
    Separator => lex::Token { kind: lex::TokenKind::Separator, .. },
  }
}