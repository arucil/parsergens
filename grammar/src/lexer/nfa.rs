use std::hash::Hash;
use std::fmt::{self, Debug};
use super::dfa::Dfa;
use super::nfa_builder::NfaBuilder;
use crate::{Map, Set};

#[derive(Debug)]
pub struct Nfa<A, P, V> {
  pub transitions: Map<(State, Option<A>), Set<State>>,
  pub alphabets: Map<State, Set<A>>,
  pub accept_states: Map<State, (P, V)>,
}

#[derive(PartialEq, Eq, Hash, Clone, Copy)]
pub struct State(pub usize);

impl<A, P, V> Nfa<A, P, V>
  where A: Eq + Hash + Copy,
        P: PartialOrd
{
  pub fn builder() -> NfaBuilder<A, P, V> {
    NfaBuilder::new()
  }
}

impl<A, P, V> Nfa<A, P, V>
  where A: Eq + Hash + Copy + Debug,
        P: PartialOrd,
        V: Clone
{
  pub fn to_dfa(&self, start: State) -> Dfa<A, V> {
    super::powerset_cons::powerset(self, start)
  }
}

impl Debug for State {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "State({})", self.0)
  }
}