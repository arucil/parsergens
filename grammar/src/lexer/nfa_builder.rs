use std::hash::Hash;
use super::nfa::*;
use super::{Map, Set};

pub struct NfaBuilder<A, P, V> {
  counter: usize,
  transitions: Map<(State, Option<A>), Set<State>>,
  alphabets: Map<State, Set<A>>,
  accept_states: Map<State, (P, V)>,
}

impl<A, P, V> NfaBuilder<A, P, V>
  where A: Eq + Hash + Copy,
        P: PartialOrd
{
  pub fn new() -> Self {
    Self {
      counter: 0,
      transitions: Map::new(),
      alphabets: Map::new(),
      accept_states: Map::new(),
    }
  }

  pub fn build(self) -> Nfa<A, P, V> {
    Nfa {
      transitions: self.transitions,
      alphabets: self.alphabets,
      accept_states: self.accept_states,
    }
  }

  pub fn state(&mut self) -> State {
    self.counter += 1;
    State(self.counter)
  }

  pub fn transition(&mut self, src: State, dest: State, c: Option<A>) {
    self.transitions.entry((src, c)).or_default().insert(dest);
    if let Some(c) = c {
      self.alphabets.entry(src).or_default().insert(c);
    }
  }

  pub fn accept(&mut self, state: State, priority: P, value: V) {
    self.accept_states.insert(state, (priority, value));
  }
}