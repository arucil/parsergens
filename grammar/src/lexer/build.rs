use std::collections::BTreeSet;
use super::super::grammar_parser::ast::*;
use super::super::grammar_parser::regex::{RegexError, RegexErrorKind};
use super::nfa::{State, Nfa};
use super::nfa_builder::NfaBuilder;
use super::{Lexer, TokenId, LexerError};
use super::util;
use crate::{Set, BiMap};

pub fn build(decls: &[&TokenDecl], skips: &[&SkipDecl]) -> Result<Lexer, LexerError> {
  if decls.is_empty() {
    return Err(LexerError::NoTokens);
  }

  validate_decls(decls, skips)?;

  let char_intervals = compute_char_intervals(decls, skips);

  let mut nfa_builder = Nfa::builder();
  let start = nfa_builder.state();
  let mut token_gen = TokenGen::default();
  let mut token_names = BiMap::new();

  for decl in decls {
    let accept = add_regex_to_nfa(
      &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
    let token = token_gen.gen();
    token_names.insert(token, decl.name.1.clone());

    nfa_builder.accept(accept, decl.pattern.source.0 .0, token);
  }

  for decl in skips {
    let accept = add_regex_to_nfa(
      &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
    let token = token_gen.gen();

    nfa_builder.accept(accept, decl.pattern.source.0 .0, token);
  }

  let nfa = nfa_builder.build();
  let dfa = nfa.to_dfa(start).minimize();

  Ok(Lexer {
    dfa,
    char_intervals,
    token_names,
  })
}

fn add_regex_to_nfa(
  builder: &mut NfaBuilder<u32, usize, TokenId>,
  regex: &Regex,
  enter: State,
  char_intervals: &[u32],
) -> State {
  match regex {
    Regex::Empty => {
      enter
    }
    Regex::Any => {
      let exit = builder.state();
      add_intervals_to_nfa(builder, enter, exit, 0, char_intervals.len() as u32);
      exit
    }
    Regex::Char(c) => {
      let exit = builder.state();
      let interval = util::find_char_interval(*c as u32, char_intervals);
      builder.transition(enter, exit, Some(interval));
      exit
    }
    Regex::CharSet(set, negated) => {
      let exit = builder.state();
      if *negated {
        let mut neg_set = (0..char_intervals.len() as u32).collect::<Set<_>>();
        for item in set {
          match item {
            CharSetItem::Char(c) => {
              let interval = util::find_char_interval(*c as u32, char_intervals);
              neg_set.remove(&interval);
            }
            CharSetItem::CharClass(class) => {
              apply_char_class(*class, char_intervals, |lower, upper| {
                for i in lower..upper {
                  neg_set.remove(&i);
                }
              });
            }
            CharSetItem::Range(a, b) => {
              let lower = util::find_char_interval(*a as u32, char_intervals);
              let upper = util::find_char_interval(*b as u32 + 1, char_intervals);
              for i in lower..upper {
                neg_set.remove(&i);
              }
            }
          }
        }

        for interval_ix in neg_set {
          builder.transition(enter, exit, Some(interval_ix));
        }
      } else {
        for item in set {
          match item {
            CharSetItem::Char(c) => {
              let interval_ix = util::find_char_interval(*c as u32, char_intervals);
              builder.transition(enter, exit, Some(interval_ix));
            }
            CharSetItem::CharClass(class) => {
              add_char_class_to_nfa(builder, enter, exit, *class, char_intervals);
            }
            CharSetItem::Range(a, b) => {
              let lower = util::find_char_interval(*a as u32, char_intervals);
              let upper = util::find_char_interval(*b as u32 + 1, char_intervals);
              add_intervals_to_nfa(builder, enter, exit, lower, upper);
            }
          }
        }
      }
      exit
    }
    Regex::CharClass(class) => {
      let exit = builder.state();
      add_char_class_to_nfa(builder, enter, exit, *class, char_intervals);
      exit
    }
    Regex::Alt(alts) => {
      let exit = builder.state();
      for alt in alts {
        let last = add_regex_to_nfa(builder, alt, enter, char_intervals);
        builder.transition(last, exit, None);
      }
      exit
    }
    Regex::Concat(items) => {
      items.iter().fold(enter, |enter, item| {
        add_regex_to_nfa(builder, item, enter, char_intervals)
      })
    }
    Regex::Optional(item) => {
      let exit = add_regex_to_nfa(builder, item, enter, char_intervals);
      builder.transition(enter, exit, None);
      exit
    }
    Regex::Many(item) => {
      let exit = builder.state();
      builder.transition(enter, exit, None);
      let temp = add_regex_to_nfa(builder, item, exit, char_intervals);
      builder.transition(temp, exit, None);
      exit
    }
    Regex::Many1(item) => {
      let temp = builder.state();
      builder.transition(enter, temp, None);
      let exit = add_regex_to_nfa(builder, item, temp, char_intervals);
      builder.transition(exit, temp, None);
      exit
    }
  }
}

fn add_char_class_to_nfa(
  builder: &mut NfaBuilder<u32, usize, TokenId>,
  enter: State,
  exit: State,
  class: CharClass,
  char_intervals: &[u32]
) {
  apply_char_class(class, char_intervals, |lower, upper| {
    add_intervals_to_nfa(builder, enter, exit, lower, upper);
  });
}

fn apply_char_class<F>(
  class: CharClass,
  char_intervals: &[u32],
  mut action: F
)
  where F: FnMut(u32, u32)
{
  for (c0, c1) in class.ranges() {
    let lower = util::find_char_interval(c0 as u32, char_intervals);
    let upper = util::find_char_interval(c1 as u32 + 1, char_intervals);
    action(lower, upper);
  }
}

fn add_intervals_to_nfa(
  builder: &mut NfaBuilder<u32, usize, TokenId>,
  enter: State,
  exit: State,
  lower_ix: u32,
  upper_ix: u32
) {
  for i in lower_ix..upper_ix {
    builder.transition(enter, exit, Some(i));
  }
}

fn compute_char_intervals(
  decls: &[&TokenDecl],
  skips: &[&SkipDecl]
) -> Vec<u32> {
  let mut char_intervals = BTreeSet::new();
  char_intervals.insert(0);

  for decl in decls {
    collect_char_intervals(&decl.pattern.regex, &mut char_intervals);
  }

  for decl in skips {
    collect_char_intervals(&decl.pattern.regex, &mut char_intervals);
  }

  char_intervals.into_iter().collect()
}

fn collect_char_intervals(regex: &Regex, intervals: &mut BTreeSet<u32>) {
  match regex {
    Regex::Any => {}
    Regex::Empty => {}
    Regex::Char(c) => {
      intervals.insert(*c as u32);
      intervals.insert(*c as u32 + 1);
    }
    Regex::CharSet(set, _) => {
      for item in set {
        match item {
          CharSetItem::Char(c) => {
            intervals.insert(*c as u32);
            intervals.insert(*c as u32 + 1);
          }
          CharSetItem::CharClass(class) => {
            collect_char_class_intervals(class, intervals);
          }
          CharSetItem::Range(a, b) => {
            intervals.insert(*a as u32);
            intervals.insert(*b as u32 + 1);
          }
        }
      }
    }
    Regex::CharClass(class) => {
      collect_char_class_intervals(class, intervals);
    }
    Regex::Alt(items) | Regex::Concat(items) => {
      for item in items {
        collect_char_intervals(item, intervals);
      }
    }
    Regex::Optional(item) | Regex::Many(item) | Regex::Many1(item) => {
      collect_char_intervals(item, intervals);
    }
  }
}

fn collect_char_class_intervals(class: &CharClass, intervals: &mut BTreeSet<u32>) {
  for (c0, c1) in class.ranges() {
    intervals.insert(c0 as u32);
    intervals.insert(c1 as u32 + 1);
  }
}

fn validate_decls(
  decls: &[&TokenDecl],
  skips: &[&SkipDecl]
) -> Result<(), LexerError> {
  for decl in decls {
    if regex_accepts_empty_string(&decl.pattern.regex) {
      return Err(LexerError::RegexError(RegexError {
        kind: RegexErrorKind::Empty,
        span: decl.pattern.source.0,
      }));
    }
  }

  for decl in skips {
    if regex_accepts_empty_string(&decl.pattern.regex) {
      return Err(LexerError::RegexError(RegexError {
        kind: RegexErrorKind::Empty,
        span: decl.pattern.source.0,
      }));
    }
  }

  Ok(())
}

fn regex_accepts_empty_string(regex: &Regex) -> bool {
  match regex {
    Regex::Empty => true,
    Regex::Alt(alts) => alts.iter().any(regex_accepts_empty_string),
    Regex::Concat(items) => items.iter().all(regex_accepts_empty_string),
    Regex::Many(_) | Regex::Optional(_) => true,
    Regex::Many1(item) => regex_accepts_empty_string(item),
    Regex::Any | Regex::Char(_) | Regex::CharClass(_) => false,
    Regex::CharSet(set, negated) => {
      if *negated {
        // TODO this needs further consideration
        false
      } else {
        char_set_accepts_empty_string(set)
      }
    }
  }
}

fn char_set_accepts_empty_string(set: &[CharSetItem]) -> bool {
  set.iter()
    .all(|item| {
      match item {
        CharSetItem::Char(_) => false,
        CharSetItem::CharClass(_) => false,
        CharSetItem::Range(a, b) => a > b,
      }
    })
}

#[derive(Default)]
struct TokenGen(u32);

impl TokenGen {
  fn gen(&mut self) -> TokenId {
    let i = self.0;
    self.0 += 1;
    TokenId(i)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use super::super::super::grammar_parser;
  use insta::assert_debug_snapshot;

  #[test]
  fn longest_match() {
    let ast = grammar_parser::parse(r#"
%token IF "if"
%token IN "int"
%token ID /[a-zA-Z_][\w_]*/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl),
        _ => None,
      })
      .collect::<Vec<_>>();

    let char_intervals = compute_char_intervals(&decls, &[]);

    let mut nfa_builder = NfaBuilder::new();
    let start = nfa_builder.state();
    let mut token_gen = TokenGen::default();

    for decl in decls {
      let accept = add_regex_to_nfa(
        &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
      let token = token_gen.gen();

      nfa_builder.accept(accept, !decl.pattern.source.0 .0, token);
    }

    let nfa = nfa_builder.build();
    let dfa = nfa.to_dfa(start);

    assert_debug_snapshot!((
      char_intervals.iter().enumerate().collect::<Vec<_>>(),
      nfa,
      dfa,
    ));
  }

  #[test]
  fn regex_repetition() {
    let ast = grammar_parser::parse(r#"
%token INT /\d+(\.\d*)?/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl),
        _ => None,
      })
      .collect::<Vec<_>>();

    let char_intervals = compute_char_intervals(&decls, &[]);

    let mut nfa_builder = NfaBuilder::new();
    let start = nfa_builder.state();
    let mut token_gen = TokenGen::default();

    for decl in decls {
      let accept = add_regex_to_nfa(
        &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
      let token = token_gen.gen();

      nfa_builder.accept(accept, !decl.pattern.source.0 .0, token);
    }

    let nfa = nfa_builder.build();
    let dfa = nfa.to_dfa(start);

    assert_debug_snapshot!((
      char_intervals.iter().enumerate().collect::<Vec<_>>(),
      nfa,
      dfa,
    ));
  }

  #[test]
  fn line_comment() {
    let ast = grammar_parser::parse(r#"
%token COMMENT /#[^\n]*/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl),
        _ => None,
      })
      .collect::<Vec<_>>();

    let char_intervals = compute_char_intervals(&decls, &[]);

    let mut nfa_builder = NfaBuilder::new();
    let start = nfa_builder.state();
    let mut token_gen = TokenGen::default();

    for decl in decls {
      let accept = add_regex_to_nfa(
        &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
      let token = token_gen.gen();

      nfa_builder.accept(accept, !decl.pattern.source.0 .0, token);
    }

    let nfa = nfa_builder.build();
    let dfa = nfa.to_dfa(start);

    assert_debug_snapshot!((
      char_intervals.iter().enumerate().collect::<Vec<_>>(),
      nfa,
      dfa,
    ));
  }
}