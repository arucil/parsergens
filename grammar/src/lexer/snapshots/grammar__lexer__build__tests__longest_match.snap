---
source: grammar/src/lexer/build.rs
expression: "(char_intervals.iter().enumerate().collect::<Vec<_>>(), nfa, dfa)"
---
(
    [
        (
            0,
            0,
        ),
        (
            1,
            48,
        ),
        (
            2,
            58,
        ),
        (
            3,
            65,
        ),
        (
            4,
            91,
        ),
        (
            5,
            95,
        ),
        (
            6,
            96,
        ),
        (
            7,
            97,
        ),
        (
            8,
            102,
        ),
        (
            9,
            103,
        ),
        (
            10,
            105,
        ),
        (
            11,
            106,
        ),
        (
            12,
            110,
        ),
        (
            13,
            111,
        ),
        (
            14,
            116,
        ),
        (
            15,
            117,
        ),
        (
            16,
            123,
        ),
    ],
    Nfa {
        transitions: {
            (
                State(0),
                Some(
                    10,
                ),
            ): {
                State(1),
                State(3),
                State(6),
            },
            (
                State(1),
                Some(
                    8,
                ),
            ): {
                State(2),
            },
            (
                State(3),
                Some(
                    12,
                ),
            ): {
                State(4),
            },
            (
                State(4),
                Some(
                    14,
                ),
            ): {
                State(5),
            },
            (
                State(0),
                Some(
                    7,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    8,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    9,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    11,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    12,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    13,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    14,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    15,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    3,
                ),
            ): {
                State(6),
            },
            (
                State(0),
                Some(
                    5,
                ),
            ): {
                State(6),
            },
            (
                State(6),
                None,
            ): {
                State(7),
            },
            (
                State(7),
                Some(
                    1,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    7,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    8,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    9,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    10,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    11,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    12,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    13,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    14,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    15,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    3,
                ),
            ): {
                State(8),
            },
            (
                State(7),
                Some(
                    5,
                ),
            ): {
                State(8),
            },
            (
                State(8),
                None,
            ): {
                State(7),
            },
        },
        alphabets: {
            State(0): {
                10,
                7,
                8,
                9,
                11,
                12,
                13,
                14,
                15,
                3,
                5,
            },
            State(1): {
                8,
            },
            State(3): {
                12,
            },
            State(4): {
                14,
            },
            State(7): {
                1,
                7,
                8,
                9,
                10,
                11,
                12,
                13,
                14,
                15,
                3,
                5,
            },
        },
        accept_states: {
            State(2): (
                18446744073709551604,
                TokenId(
                    0,
                ),
            ),
            State(5): (
                18446744073709551589,
                TokenId(
                    1,
                ),
            ),
            State(7): (
                18446744073709551573,
                TokenId(
                    2,
                ),
            ),
        },
    },
    Dfa {
        start: 0,
        transitions: {
            (
                State(0),
                10,
            ): State(1),
            (
                State(0),
                7,
            ): State(2),
            (
                State(0),
                8,
            ): State(2),
            (
                State(0),
                9,
            ): State(2),
            (
                State(0),
                11,
            ): State(2),
            (
                State(0),
                12,
            ): State(2),
            (
                State(0),
                13,
            ): State(2),
            (
                State(0),
                14,
            ): State(2),
            (
                State(0),
                15,
            ): State(2),
            (
                State(0),
                3,
            ): State(2),
            (
                State(0),
                5,
            ): State(2),
            (
                State(1),
                8,
            ): State(3),
            (
                State(1),
                12,
            ): State(4),
            (
                State(1),
                1,
            ): State(5),
            (
                State(1),
                7,
            ): State(5),
            (
                State(1),
                9,
            ): State(5),
            (
                State(1),
                10,
            ): State(5),
            (
                State(1),
                11,
            ): State(5),
            (
                State(1),
                13,
            ): State(5),
            (
                State(1),
                14,
            ): State(5),
            (
                State(1),
                15,
            ): State(5),
            (
                State(1),
                3,
            ): State(5),
            (
                State(1),
                5,
            ): State(5),
            (
                State(2),
                1,
            ): State(5),
            (
                State(2),
                7,
            ): State(5),
            (
                State(2),
                8,
            ): State(5),
            (
                State(2),
                9,
            ): State(5),
            (
                State(2),
                10,
            ): State(5),
            (
                State(2),
                11,
            ): State(5),
            (
                State(2),
                12,
            ): State(5),
            (
                State(2),
                13,
            ): State(5),
            (
                State(2),
                14,
            ): State(5),
            (
                State(2),
                15,
            ): State(5),
            (
                State(2),
                3,
            ): State(5),
            (
                State(2),
                5,
            ): State(5),
            (
                State(3),
                1,
            ): State(5),
            (
                State(3),
                7,
            ): State(5),
            (
                State(3),
                8,
            ): State(5),
            (
                State(3),
                9,
            ): State(5),
            (
                State(3),
                10,
            ): State(5),
            (
                State(3),
                11,
            ): State(5),
            (
                State(3),
                12,
            ): State(5),
            (
                State(3),
                13,
            ): State(5),
            (
                State(3),
                14,
            ): State(5),
            (
                State(3),
                15,
            ): State(5),
            (
                State(3),
                3,
            ): State(5),
            (
                State(3),
                5,
            ): State(5),
            (
                State(4),
                14,
            ): State(6),
            (
                State(4),
                1,
            ): State(5),
            (
                State(4),
                7,
            ): State(5),
            (
                State(4),
                8,
            ): State(5),
            (
                State(4),
                9,
            ): State(5),
            (
                State(4),
                10,
            ): State(5),
            (
                State(4),
                11,
            ): State(5),
            (
                State(4),
                12,
            ): State(5),
            (
                State(4),
                13,
            ): State(5),
            (
                State(4),
                15,
            ): State(5),
            (
                State(4),
                3,
            ): State(5),
            (
                State(4),
                5,
            ): State(5),
            (
                State(5),
                1,
            ): State(5),
            (
                State(5),
                7,
            ): State(5),
            (
                State(5),
                8,
            ): State(5),
            (
                State(5),
                9,
            ): State(5),
            (
                State(5),
                10,
            ): State(5),
            (
                State(5),
                11,
            ): State(5),
            (
                State(5),
                12,
            ): State(5),
            (
                State(5),
                13,
            ): State(5),
            (
                State(5),
                14,
            ): State(5),
            (
                State(5),
                15,
            ): State(5),
            (
                State(5),
                3,
            ): State(5),
            (
                State(5),
                5,
            ): State(5),
            (
                State(6),
                1,
            ): State(5),
            (
                State(6),
                7,
            ): State(5),
            (
                State(6),
                8,
            ): State(5),
            (
                State(6),
                9,
            ): State(5),
            (
                State(6),
                10,
            ): State(5),
            (
                State(6),
                11,
            ): State(5),
            (
                State(6),
                12,
            ): State(5),
            (
                State(6),
                13,
            ): State(5),
            (
                State(6),
                14,
            ): State(5),
            (
                State(6),
                15,
            ): State(5),
            (
                State(6),
                3,
            ): State(5),
            (
                State(6),
                5,
            ): State(5),
        },
        accept_states: {
            State(1): TokenId(
                2,
            ),
            State(2): TokenId(
                2,
            ),
            State(3): TokenId(
                0,
            ),
            State(4): TokenId(
                2,
            ),
            State(5): TokenId(
                2,
            ),
            State(6): TokenId(
                1,
            ),
        },
    },
)
